File: bcs-cc/storage/models/cluster.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
 * Copyright (C) 2017-2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package models

import (
	"errors"
	"fmt"
	"time"

	"bcs_cc/config"
	"bcs_cc/storage"
	"bcs_cc/utils"

	"github.com/jinzhu/gorm"
)

// Cluster Model : cluster info
//go:generate goqueryset -in ${GOFILE} -out qs_${GOFILE}
// gen:qs
type Cluster struct {
	Model
	Name              string     `json:"name" gorm:"size:64;unique_index:uix_project_id_name"`
	Creator           string     `json:"creator" gorm:"size:32"`
	Description       string     `json:"description" sql:"size:128"`
	ProjectID         string     `json:"project_id" gorm:"size:32;index;unique_index:uix_project_id_name"`
	RelatedProjects   string     `json:"related_projects" sql:"type:text"`
	ClusterID         string     `json:"cluster_id" gorm:"size:64;unique_index"`
	ClusterNum        int64      `json:"cluster_num" gorm:"unique"`
	Status            string     `json:"status" gorm:"size:64"`     // 初始化的状态 uninitialized, initializing, initialized, initialize_failed
	Disabled          bool       `json:"disabled"`                  // 是否禁用掉了，default false
	Type              string     `json:"type" gorm:"size:8"`        // mesos,k8s
	Environment       string     `json:"environment" gorm:"size:8"` // stag,debug,prod
	AreaID            int        `json:"area_id"`
	ConfigSvrCount    int        `json:"config_svr_count"`
	MasterCount       int        `json:"master_count"`
	NodeCount         int        `json:"node_count"`
	IPResourceTotal   int        `json:"ip_resource_total"`
	IPResourceUsed    int        `json:"ip_resource_used"`
	Artifactory       string     `json:"artifactory" gorm:"size:256"`
	TotalMem          float64    `json:"total_mem"`
	RemainMem         float64    `json:"remain_mem"`
	TotalCPU          float64    `json:"total_cpu"`
	RemainCPU         float64    `json:"remain_cpu"`
	TotalDisk         float64    `json:"total_disk"`
	RemainDisk        float64    `json:"remain_disk"`
	CapacityUpdatedAt *time.Time `json:"capacity_updated_at"`
	NotNeedNAT        bool       `json:"not_need_nat" gorm:"default:false"`
	ExtraClusterID    string     `json:"extra_cluster_id" gorm:"size:64"`
}

// ClusterFilterParams :
// note: include the embedded struct
type ClusterFilterParams struct {
	ProjectID     string
	InnerIP       string
	ClusterIDList []string
	Name          string
	DesireAllData string
	Limit         int
	Offset        int
	ClusterType   string
	FilterStatus  []string
	ExcludeStatus []string
}

// UpdateClusterDataJSON :
type UpdateClusterDataJSON struct {
	ProjectID         string    `json: "project_id"`
	ClusterID         string    `json:"cluster_id"`
	Name              string    `json:"name"`
	Description       *string   `json:"description"`
	Status            string    `json:"status"`
	Disabled          *bool     `json:"disabled"`
	Environment       string    `json:"environment"`
	AreaID            *int      `json:"area_id"`
	ConfigSvrCount    *int      `json:"config_svr_count"`
	MasterCount       *int      `json:"master_count"`
	NodeCount         *int      `json:"node_count"`
	Artifactory       string    `json:"artifactory"`
	TotalMem          *float64  `json:"total_mem"`
	RemainMem         *float64  `json:"remain_mem"`
	TotalCPU          *float64  `json:"total_cpu"`
	RemainCPU         *float64  `json:"remain_cpu"`
	TotalDisk         *float64  `json:"total_disk"`
	RemainDisk        *float64  `json:"remain_disk"`
	CapacityUpdatedAt string    `json:"capacity_updated_at"`
	IPResourceTotal   *int      `json:"ip_resource_total"`
	IPResourceUsed    *int      `json:"ip_resource_used"`
	NeedNAT           *bool     `json:"need_nat"`
	AgentList         *[]string `json:"agent_list"`
	RelatedProjects   *[]string `json:"related_projects"`
	ExtraClusterID    *string   `json:"extra_cluster_id"`
}

func clusterListBase(filter ClusterFilterParams) (clusterList []Cluster, err error) {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	// filter
	qs = qs.ClusterListByProjectID(
		filter.ProjectID,
	).ClusterIDInWithoutError(
		filter.ClusterIDList...,
	).ClusterTypeEqWithoutExist(
		filter.ClusterType,
	).ClusterStatusNotInWithoutError(
		filter.ExcludeStatus...,
	).ClusterStatusInWithoutError(filter.FilterStatus...)
	if err := qs.All(&clusterList); err != nil {
		return nil, err
	}
	return clusterList, nil
}

// ClusterWithoutRemoved :
func ClusterWithoutRemoved(excludeStatus []string) ([]Cluster, error) {
	return clusterListBase(ClusterFilterParams{ExcludeStatus: excludeStatus})
}

// ClusterList : fetch cluster list by params
func ClusterList(clusterType string, filterStatus []string) ([]Cluster, error) {
	return clusterListBase(ClusterFilterParams{
		ClusterType: clusterType, FilterStatus: filterStatus,
	})
}

// FetchPublicClusterList : fetch public cluster info
func FetchPublicClusterList(projectID string) (data []Cluster, err error) {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	if err := qs.RelatedProjectsLike(projectID).All(&data); err != nil {
		return nil, err
	}
	return data, nil
}

// FetchPublicClusterIDList : get public cluster id by project id
func FetchPublicClusterIDList(projectID string) (clusterIDList []string, err error) {
	clusterList, err := FetchPublicClusterList(projectID)
	if err != nil {
		return nil, err
	}
	for _, clusterInfo := range clusterList {
		clusterIDList = append(clusterIDList, clusterInfo.ClusterID)
	}
	return clusterIDList, nil
}

// ValidateProjectCluster : cluster of project is valid
func ValidateProjectCluster(projectID string, clusterID string) error {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	if err := qs.ClusterListByProjectID(
		projectID,
	).ClusterIDEq(clusterID).One(&Cluster{}); err != nil {
		return errors.New("cluster not found in project")
	}
	return nil
}

// RetriveRecord : get a record
func (cluster *Cluster) RetriveRecord() error {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	if err := qs.ClusterIDEq(cluster.ClusterID).One(cluster); err != nil {
		return err
	}
	return nil
}

// RetriveRecordByProjectIDClusterID : get a record
func (cluster *Cluster) RetriveRecordByProjectIDClusterID() error {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	if err := qs.EnvironmentIn(
		config.GlobalConfigurations.AvailableEnvironmentFlags...,
	).ClusterListByProjectID(
		cluster.ProjectID,
	).ClusterIDEq(cluster.ClusterID).One(cluster); err != nil {
		return err
	}
	return nil
}

// FetchPrivateClusterByProjectID : get project private cluster info
func FetchPrivateClusterByProjectID(projectID string) (data []Cluster, err error) {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	if err := qs.ProjectIDEq(projectID).All(&data); err != nil {
		return nil, err
	}
	return data, nil
}

// FetchPrivateClusterIDByProjectID : get project private cluster id
func FetchPrivateClusterIDByProjectID(projectID string) (clusterIDList []string, err error) {
	clusterList, err := FetchPrivateClusterByProjectID(projectID)
	if err != nil {
		return nil, err
	}
	for _, clusterInfo := range clusterList {
		clusterIDList = append(clusterIDList, clusterInfo.ClusterID)
	}
	return clusterIDList, nil
}

func (qs *ClusterQuerySet) queryClusterWithDesireAll(filter *ClusterFilterParams) (clusterList []Cluster, err error) {
	if err := qs.ClusterListByProjectID(filter.ProjectID).All(&clusterList); err != nil {
		return nil, err
	}
	return clusterList, nil
}

func getClusterIDList(clusterList []Cluster) (clusterIDList []string) {
	for _, clusterInfo := range clusterList {
		clusterIDList = append(clusterIDList, clusterInfo.ClusterID)
	}
	return clusterIDList
}

// queryClusterIDWithDesireAll : get project all cluster id
func (filter *ClusterFilterParams) clusterIDListWithAll(qs ClusterQuerySet) (clusterIDList []string, err error) {
	var clusterList []Cluster
	if err := qs.ClusterListByProjectID(filter.ProjectID).All(&clusterList); err != nil {
		return nil, err
	}

	return getClusterIDList(clusterList), nil
}

// get cluster id list by filter params
func (filter *ClusterFilterParams) clusterIDList(qs ClusterQuerySet) (clusterIDList []string, err error) {
	clusterID := ""
	// get cluster id by master ip
	if filter.InnerIP != "" {
		master := &ManagerMaster{InnerIP: filter.InnerIP, ProjectID: filter.ProjectID}
		if err := master.RetriveRecord(); err != nil {
			clusterID = NotExistClusterID
		} else {
			clusterID = master.ClusterID
		}
	}
	var clusterList []Cluster
	tempQuerySet := qs.ClusterIDEqWithEmpty(clusterID).NameEqWithEmpty(
		filter.Name,
	).ClusterListByProjectID(filter.ProjectID)
	if err := tempQuerySet.All(&clusterList); err != nil {
		return nil, err
	}
	return getClusterIDList(clusterList), nil
}

// ClusterListInfo : get cluster list
func (filter *ClusterFilterParams) ClusterListInfo() (data []Cluster, count int, err error) {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	var clusterIDList []string
	if filter.DesireAllData != "1" {
		clusterIDList, err = filter.clusterIDList(qs)
		if err != nil {
			return nil, 0, err
		}
	} else {
		clusterIDList, err = filter.clusterIDListWithAll(qs)
		if err != nil {
			return nil, 0, err
		}
	}
	// if length of cluster id list is 0,  return nil, 0, nil
	if len(clusterIDList) == 0 {
		return nil, 0, nil
	}
	// final filter cluster
	qs = qs.ClusterIDIn(clusterIDList...)
	count, err = qs.Count()
	if err != nil {
		return nil, 0, err
	}
	// limit and offset for cluster pagination
	if filter.DesireAllData != "1" {
		qs = qs.Limit(filter.Limit).Offset(filter.Offset)
	}
	if err := qs.All(&data); err != nil {
		return nil, 0, err
	}
	return data, count, nil
}

// ClusterListByID : get cluster info by cluster id list
func (filter *ClusterFilterParams) ClusterListByID() (data []Cluster, err error) {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	var clusterList []Cluster
	qs = qs.ClusterListByProjectID(filter.ProjectID).ClusterIDIn(filter.ClusterIDList...)
	if err := qs.All(&clusterList); err != nil {
		return nil, err
	}
	return clusterList, nil
}

// DeleteRecord : delete a cluster with master by db transaction
func (cluster *Cluster) DeleteRecord() error {
	db := storage.GetDefaultSession().DB
	tx := db.Begin()
	// delete master records
	masterQuerySet := NewManagerMasterQuerySet(tx)
	if err := masterQuerySet.ClusterIDEq(cluster.ClusterID).DeleteUnscoped(); err != nil {
		tx.Rollback()
		return err
	}
	// 修改集群名称
	clusterQuerySet := NewClusterQuerySet(tx)
	clusterQuerySet = clusterQuerySet.ClusterIDEq(cluster.ClusterID)
	updater := clusterQuerySet.GetUpdater()
	if err := updater.SetName("deleted:" + cluster.ClusterID).Update(); err != nil {
		tx.Rollback()
		return err
	}
	// delete cluster record
	if err := clusterQuerySet.Delete(); err != nil {
		tx.Rollback()
		return err
	}
	tx.Commit()
	return nil
}

// GetMaxClusterNum : get the max cluster id number
func GetMaxClusterNum(clusterType string, clusterEnv string) (clusterNum int, err error) {
	db := storage.GetDefaultSession().DB
	sql := fmt.Sprintf("select max(cluster_num) as max_num from clusters where type='%v' and environment='%v'", clusterType, clusterEnv)
	var maxInst struct{ MaxNum int }
	if err := db.Raw(sql).Scan(&maxInst).Error; err != nil {
		return 0, err
	}
	return maxInst.MaxNum + 1, nil
}

// CreateRecordWithMaster :
func (cluster *Cluster) CreateRecordWithMaster(masterList []ManagerMaster) error {
	db := storage.GetDefaultSession().DB
	tx := db.Begin()
	if err := cluster.CreateRecordWithDB(tx); err != nil {
		tx.Rollback()
		return err
	}
	for _, master := range masterList {
		master.ProjectID = cluster.ProjectID
		master.ClusterID = cluster.ClusterID
		if err := master.CreateRecordWithDB(tx); err != nil {
			tx.Rollback()
			return err
		}
	}
	tx.Commit()
	if err := cluster.RetriveRecord(); err != nil {
		return err
	}
	return nil
}

// CreateRecord : create a cluster info
func (cluster *Cluster) CreateRecord() error {
	return nil
}

// CreateRecordWithDB :
func (cluster *Cluster) CreateRecordWithDB(db *gorm.DB) error {
	if err := cluster.Create(db); err != nil {
		return err
	}
	return nil
}

// UpdateRecord :
func (updaterData UpdateClusterDataJSON) UpdateRecord() error {
	currTime, err := time.ParseInLocation(
		"2006-01-02 15:04:05", updaterData.CapacityUpdatedAt, time.Local,
	)
	if err != nil {
		currTime = time.Now()
	}
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	qs = qs.ClusterListByProjectID(updaterData.ProjectID).ClusterIDEq(updaterData.ClusterID)
	var cluster Cluster
	if err := qs.One(&cluster); err != nil {
		return err
	}
	clusterUpdater := qs.GetUpdater()
	clusterUpdater = clusterUpdater.SetNameWithoutEmpty(
		updaterData.Name).SetDescriptionWithoutNil(
		updaterData.Description).SetDisableWithoutNil(
		updaterData.Disabled).SetAreaIDWithoutNil(
		updaterData.AreaID).SetTotalCPUWithoutNil(
		updaterData.TotalCPU).SetRemainCPUWithoutNil(
		updaterData.RemainCPU).SetTotalDiskWithoutNil(
		updaterData.TotalDisk).SetRemainDiskWithoutNil(
		updaterData.RemainDisk).SetTotalMemWithoutNil(
		updaterData.TotalMem).SetRemainMemWithoutNil(
		updaterData.RemainMem).SetStatusWithoutEmpty(
		updaterData.Status).SetCapacityUpdatedAt(
		&currTime).SetIPResourceTotalWithoutNil(
		updaterData.IPResourceTotal).SetIPResourceUsedWithoutNil(
		updaterData.IPResourceUsed).SetNotNeedNATWithoutNil(
		updaterData.NeedNAT).SetExtraClusterIDWithoutNil(updaterData.ExtraClusterID)
	updateRelatedProjects := ""
	relatedProjects := updaterData.RelatedProjects
	if relatedProjects != nil {
		projectList := utils.StringSplit(cluster.RelatedProjects, ";")
		for _, id := range *relatedProjects {
			if utils.StringInSlice(id, projectList) {
				continue
			}
			if id != "" {
				projectList = append(projectList, id)
			}
		}
		updateRelatedProjects = utils.ListJoinWithString(&projectList, ";")
	}
	clusterUpdater = clusterUpdater.SetRelatedProjectsWithoutEmpty(updateRelatedProjects)
	if err := clusterUpdater.Update(); err != nil {
		return err
	}
	return nil
}

// ClusterListWithoutRemoved : get cluster list by project info
func ClusterListWithoutRemoved(status []string) (data []Cluster, err error) {
	qs := NewClusterQuerySet(storage.GetDefaultSession().DB)
	if err := qs.StatusNotIn(status...).All(&data); err != nil {
		return nil, err
	}
	return data, nil
}

File: bcs-app/backend/components/bcs_api/cluster.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
from typing import Dict, List

from attr import asdict, dataclass
from django.conf import settings

from ..base import BaseHttpClient, BkApiClient, ComponentAuth, response_handler
from . import BcsApiAuth
from .constants import BCS_API_VERSION, ClusterManageType, ManageType


class BcsApiGatewayConfig:
    def __init__(self, host: str, env: str):
        self.host = host

        _cluster_api_url_prefix = f"{host}/{env}/{BCS_API_VERSION}/clustermanager/v1"
        # cluster manager中集群相关接口url
        self.add_cluster_url = f"{_cluster_api_url_prefix}/cluster/{{cluster_id}}"
        self.delete_cluster_url = f"{_cluster_api_url_prefix}/cluster/{{cluster_id}}"
        self.update_cluster_url = f"{_cluster_api_url_prefix}/cluster/{{cluster_id}}"
        self.add_nodes_url = f"{_cluster_api_url_prefix}/cluster/{{cluster_id}}/node"
        self.delete_nodes_url = f"{_cluster_api_url_prefix}/cluster/{{cluster_id}}/node"
        self.query_task_url = f"{_cluster_api_url_prefix}/task/{{task_id}}"


@dataclass
class CloudClusterConfig:
    """集群初始化需要的基本配置
    region: 区域
    manageType: 集群管理类型，公有云时生效，MANAGED_CLUSTER(云上托管集群)，INDEPENDENT_CLUSTER(独立集群，自行维护)
    master: 集群master
    vpcID: vpc ID
    cloudID: 所属云ID
    nodes: 集群的节点
    networkSettings: 网络配置
    clusterBasicSettings: 集群的基本配置
    clusterAdvanceSettings: 集群的高级配置
    nodeSettings: 节点配置信息
    systemReinstall: 是否重装master节点的系统，机器被托管情况下有效
    initLoginPassword: 重装master节点的系统时，初始化password，机器被托管情况下有效
    status: 状态
    """

    region: str
    master: List
    manageType: str = ManageType.MANAGED_CLUSTER
    vpcID: str = ""
    cloudID: str = ""
    nodes: List = []
    networkSettings: Dict = {}
    clusterBasicSettings: Dict = {}
    clusterAdvanceSettings: Dict = {}
    nodeSettings: Dict = {}
    systemReinstall: bool = False
    initLoginPassword: str = ""
    status: str = ""


@dataclass
class BcsBasicConfig:
    """集群初始化时，BCS需要的基本配置"""

    projectID: str
    businessID: str
    clusterID: str
    clusterName: str
    provider: str
    environment: str  # 集群环境，如prod、debug、test
    engineType: str  # 引擎类型，如k8s
    clusterType: str = ClusterManageType.SINGLE
    isExclusive: bool = False  # 是否为独占集群
    federationClusterID: str = ""  # 联邦集群ID
    labels: Dict = {}
    onlyCreateInfo: bool = False
    bcsAddons: Dict = {}
    extraAddons: Dict = {}


@dataclass
class ClusterConfig:
    """集群配置"""

    creator: str
    bcs_basic_config: BcsBasicConfig
    cloud_cluster_config: CloudClusterConfig


@dataclass
class UpdatedClusterConfig:
    """更新的集群配置选项"""

    projectID: str
    clusterID: str
    updater: str
    status: str
    clusterName: str = ""
    labels: Dict = {}


@dataclass
class NodeConfig:
    nodes: List[str]
    nodeGroupID: str = ""
    onlyCreateInfo: bool = False
    initLoginPassword: str = ""


class BcsClusterApiClient(BkApiClient):
    """访问 BCS 集群 API 服务的 Client 对象
    :param auth: 包含校验信息的对象
    """

    def __init__(self, auth: ComponentAuth):
        self._config = BcsApiGatewayConfig(host=settings.BCS_API_GW_DOMAIN, env=settings.BCS_API_GW_ENV)
        self._client = BaseHttpClient(BcsApiAuth(auth.access_token))

    def add_cluster(self, cluster_config: ClusterConfig) -> Dict:
        """添加集群
        :param creator: 创建者
        :param cluster_config: 集群初始化的配置，包含bcs平台需要的信息和集群初始化需要的信息
        :return: 返回初始化的集群的信息，任务信息，格式: {"data": {}, "task": {}}
        """
        # 组装参数
        cluster_config_data = asdict(cluster_config.cloud_cluster_config)
        cluster_config_data.update(asdict(cluster_config.bcs_basic_config), creator=cluster_config.creator)
        # 下发初始化配置
        url = self._config.add_cluster_url.format(cluster_id=cluster_config_data["clusterID"])
        return self._client.request_json("POST", url, json=cluster_config_data)

    @response_handler()
    def update_cluster(self, cluster_config: UpdatedClusterConfig) -> Dict:
        """更新集群
        :param cluster_config: 更新集群的配置
        :return: 返回集群的配置
        """
        # 组装参数，并去掉为None的属性
        cluster_config_data = {k: v for k, v in asdict(cluster_config).items() if v is not None}
        # 下发更新的配置
        url = self._config.update_cluster_url.format(cluster_id=cluster_config_data["clusterID"])
        return self._client.request_json("PUT", url, json=cluster_config_data)

    def delete_cluster(
        self,
        cluster_id: str,
        is_force: bool = False,
        is_clean_resource: bool = True,
        only_delete_info: bool = False,
    ) -> Dict:
        """删除集群
        :param cluster_id: 集群ID
        :param is_force: 是否强制删除，默认为False
        :param is_clean_resource: 强制删除有效时，是否清理机器上部署的资源
        :param only_delete_info: 是否仅删除记录的信息，默认为False
        :return: 返回删除的集群信息，任务信息
        """
        url = self._config.delete_cluster_url.format(cluster_id=cluster_id)
        params = {"isForced": is_force, "resourceClean": is_clean_resource, "onlyDeleteInfo": only_delete_info}
        return self._client.request_json("DELETE", url, params=params)

    @response_handler()
    def add_nodes(self, cluster_id: str, node_config: NodeConfig) -> Dict:
        """添加节点
        :param cluster_id: 集群ID
        :param node_config: 节点配置
        :return: 返回节点的数据，包含任务ID
        """
        url = self._config.add_nodes_url.format(cluster_id=cluster_id)
        data = asdict(node_config)
        data["clusterID"] = cluster_id
        return self._client.request_json("POST", url, json=data)

    @response_handler()
    def delete_nodes(
        self, cluster_id: str, nodes: List[str], delete_mode: str = "RETAIN", is_force: bool = False
    ) -> Dict:
        """删除节点
        :param cluster_id: 集群ID
        :param nodes: 节点列表
        :param delete_mode: 删除模式，RETAIN(移除集群，但是保留主机)，TERMINATE(只支持按量计费的机器)
        :param is_force: 是否强制删除
        :return: 返回删除的节点信息，包含任务ID
        """
        url = self._config.delete_nodes_url.format(cluster_id=cluster_id)
        data = {"clusterID": cluster_id, "nodes": nodes, "deleteMode": delete_mode, "isForce": is_force}
        return self._client.request_json("DELETE", url, json=data)

    def query_task(self, task_id: str) -> Dict:
        """查询任务状态
        :param task_id: 任务ID
        :return: 返回任务的执行详情
        """
        url = self._config.query_task_url.format(task_id=task_id)
        return self._client.request_json("GET", url)
